### PR TITLE
JBS-222 : snapshot 404 error for snapshot delete

### PR DESCRIPTION
**Pull request by rjain-iitr**

### CODE REVIEW

File: ec2api/api/snapshot.py [hunk 1]
```diff
@@ -59,8 +59,11 @@ def delete_snapshot(context, snapshot_id):
     try:
         cinder.backups.delete(snapshot_id)
     except cinder_exception.NotFound:
-        pass
-    os_snapshot=cinder.backups.get(snapshot_id)
+        raise exception.InvalidSnapshotNotFound(id=snapshot_id)
+    except cinder_exception.BadRequest:
+        os_snapshot=cinder.backups.get(snapshot_id)
+        if os_snapshot.status=="deleting" or os_snapshot.status=="deleted" :
```
Comment (shishirng) [inline ec2api/api/snapshot.py]: if status="deleted" then, they wont be part of the list right?

File: ec2api/api/snapshot.py [hunk 2]
```diff
@@ -147,6 +158,7 @@ def _format_snapshot_no_detail(context, os_snapshot):
     status_map = {'creating': 'pending',
                   'available': 'completed',
                   'deleting': None,
+                  'deleted': None,
```
Comment (shishirng) [inline ec2api/api/snapshot.py]: deleted snapshots wont show up. no need to add this filter

File: ec2api/api/snapshot.py [hunk 3]
```diff
@@ -168,6 +180,7 @@ def _format_snapshot(context, os_snapshot):
     status_map = {'creating': 'pending',
                   'available': 'completed',
                   'deleting': None,
+		  'deleted':None,
```
Comment (shishirng) [inline ec2api/api/snapshot.py]: deleted snapshots wont show up. no need to add this filter

File: ec2api/api/snapshot.py [hunk 4]
```diff
@@ -73,11 +76,19 @@ def describe(self, context, ids=None, detail=True, max_results=None, next_token=
         os_items = self.get_os_items(ids, max_results, next_token, detail)
         formatted_items = []
         
-        if ((ids is not None) and isinstance(ids, list)) :
+	if ((ids is not None) and not isinstance(ids, list)):
+            for os_item in os_items:
+                   if os_item.status=="deleting" or os_item.status=="deleted" :
+                       raise exception.InvalidSnapshotNotFound(id=os_item.id)
```
Comment (shishirng) [inline ec2api/api/snapshot.py]: we shouldnt error out (404) even if 1 out all the returned snapshots are in "deleting" state
only if all are either not found or are in deleting then error out

File: ec2api/api/snapshot.py (hunk 2 above)
Comment (rjain-iitr) [inline ec2api/api/snapshot.py]: was seeing deleted in the list on devtest foe one of the instance marked deleted manually through db


<DescribeSnapshotsResponse xmlns="http://compute.jiocloudservices.com/doc/2016-03-01/">
  <requestId>req-c0ac5d63-78</requestId>
  <snapshotSet>
    <item>
      <status>deleted</status>
      <snapshotId>35c3f23c-a32e-4b47-8fb8-5c025364a99a</snapshotId>
      <volumeSize>8</volumeSize>
      <volumeId>9e373c98-eecf-4ca3-a468-3c9c21b88a0c</volumeId>
      <startTime>2016-04-14T09:55:13.000000</startTime>
    </item>
    <item>
      <status>error</status>
      <snapshotId>5ce16158-d3d0-49b5-8e65-4b868e23c78d</snapshotId>
      <volumeSize>5</volumeSize>
      <volumeId>e23d309f-fe70-4307-a6a6-8134b7aa7ce0</volumeId>
      <startTime>2016-03-20T06:50:09.000000</startTime>
    </item>
    <item>
      <status>completed</status>
      <snapshotId>93783dfb-e18a-4ee8-bfc2-8bbdbb03ff1c</snapshotId>
      <volumeSize>8</volumeSize>
      <volumeId>9e373c98-eecf-4ca3-a468-3c9c21b88a0c</volumeId>
      <startTime>2016-04-14T06:44:43.000000</startTime>
    </item>
    <item>
      <status>completed</status>
      <snapshotId>f4682c4a-016c-407f-a837-e191f078c586</snapshotId>
      <volumeSize>5</volumeSize>
      <volumeId>c6409fed-73df-4e95-8ab7-58ff79893d9d</volumeId>
      <startTime>2016-04-11T12:38:26.000000</startTime>
    </item>
  </snapshotSet>
</DescribeSnapshotsResponse>

File: ec2api/api/snapshot.py (hunk 4 above)
Comment (rjain-iitr) [inline ec2api/api/snapshot.py]: Need to know the extact expect behaviour in diferent senarios to completly do this

File: ec2api/api/snapshot.py (hunk 3 above)
Comment (rjain-iitr) [inline ec2api/api/snapshot.py]: it will just filter the deleted snapshots here

File: ec2api/api/snapshot.py [hunk 5]
```diff
@@ -73,11 +80,25 @@ def describe(self, context, ids=None, detail=True, max_results=None, next_token=
...
+
+        if ((ids is not None) and isinstance(ids, list)):
             self.ids = set(ids or [])
+	    idlist=[]
+            for os_item in os_items:
+		idlist.append(os_item.id)
+            for _id in self.ids:
```
Comment (aggarwalchirag) [inline ec2api/api/snapshot.py]: As discussed, lets use count logic here.

**Comment by aggarwalchirag** [comment]
Good to be merged. Please also list out the testcases that you executed.

**Comment by rjain-iitr** [comment]
Test cases :
1)describeSnapshots with wrong id
2)describeSnapshots with correct id
3)describeSnapshots with one wrong id in list
4)describeSnapshots with one correct id and one wrond id in list
5)describeSnapshots with same id twice in list
6) describeSnapshots with 2 correct ids in list
7)deleteSnapshot with incorrectId

**Pull request merged**